### PR TITLE
xds: move e2e tests into grpc/test/xds directory

### LINKED ISSUE #4829: xds: move e2e tests out of xds/internal
xDS e2e tests should be moved to `grpc/test` where other e2e tests live. This will ensure that we don't inadvertently use some xDS features we are not supposed to use in a e2e test.

### PR DESCRIPTION
**Pull request by easwars**
This change helps to ensure that we don't inadvertently use some xDS features we are not supposed to use in a e2e test.

Fixes #4829

RELEASE NOTES: none

### CODE REVIEW

File: xds/xds.go [hunk 1]
```diff
@@ -74,22 +77,69 @@ func init() {
 		v3statusgrpc.RegisterClientStatusDiscoveryServiceServer(grpcServer, csdss)
 		return csdss.Close, nil
 	})
+
+	internal.NewXDSResolverWithConfigForTesting = func(bootstrapConfig []byte) (resolver.Builder, error) {
```
Comment (dfawley) [inline xds/xds.go:81]: Move this to the xdsresolver package?  And can it be just:

```go
internal.NewXDSResolverWithConfigForTesting = xdsresolver.NewBuilderForTesting
```

File: internal/testutils/xds/e2e/setup_certs.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
```
Comment (dfawley) [inline internal/testutils/xds/e2e/setup_certs.go:3]: :cry:  why didn't it detect these as moves?  Is it possible to fix?

File: test/xds/xds_client_affinity_test.go [hunk 3]
```diff
@@ -1,6 +1,3 @@
-//go:build !386
```
Comment (dfawley) [inline test/xds/xds_client_affinity_test.go:1]: Why was this here before?

File: xds/xds.go [hunk 4]
```diff
@@ -74,22 +77,69 @@ func init() {
...
 //
-// This function should ONLY be used for testing and may not work with some
-// other features, including the CSDS service.
+// This function should ONLY be used for testing purposes.
 //
 // Experimental
 //
-// Notice: This API is EXPERIMENTAL and may be changed or removed in a
-// later release.
-func NewXDSResolverWithConfigForTesting(bootstrapConfig []byte) (resolver.Builder, error) {
```
Comment (dfawley) [inline xds/xds.go:93]: Why was this moved to `internal` but these other, temporary, RLS/RBAC testing functions were added here?

File: xds/xds.go (hunk 1 above)
Comment (easwars) [reply xds/xds.go:81]: Done.

File: test/xds/xds_client_affinity_test.go (hunk 3 above)
Comment (easwars) [reply test/xds/xds_client_affinity_test.go:1]: I can't remember exactly. But the one issue I remember is with go-control-plane, where they had an int64 in a struct which was accessed atomically, but wasn't aligned on 64-bit boundaries. I fixed it, but we had to wait a while for a release. So, I think we started adding these and over time, I'm guessing newer files just mimicked the existing ones.

File: internal/testutils/xds/e2e/setup_certs.go (hunk 2 above)
Comment (easwars) [reply internal/testutils/xds/e2e/setup_certs.go:3]: I initially did this as a `git mv`. But then realized that we had some commonly used helper functions like this which were defined in some _test.go files (probably the file where it was first defined). I thought it would be better to move these out into separate files which only have related helpers.

File: xds/xds.go (hunk 4 above)
Comment (easwars) [reply xds/xds.go:93]: This function is used from a couple of places: `internal/testutils/xds/e2e` and `test/xds` while the other temporary register/unregister funcs are only used from `test/xds`. There are still some tests in `xds/internal/` which use the `internal/testutils/xds/e2e` package. So, there were some import loops. For example: top-level xds package --> xds/csds --> internal/testutils/xds/e2e --> top-level-xds-package (for the NewXDSResolverWithConfigForTesting func).

I will try to move those tests out after this PR.

**Review by easwars — COMMENTED** [review]
Thanks for the review.

File: internal/testutils/xds/e2e/setup_certs.go (hunk 2 above)
Comment (dfawley) [reply internal/testutils/xds/e2e/setup_certs.go:3]: Ah, ok.. I think in that case, since it's a brand new file, the (C) date should be 2022.

File: xds/xds.go (hunk 4 above)
Comment (dfawley) [reply xds/xds.go:93]: Couldn't these also live in `internal`, though, just like the new `NewXDSResolverWithConfigForTesting`?  That should avoid import loops, too, right?

Comment (easwars) [reply xds/xds.go:93]: The import loop was only with `NewXDSResolverWithConfigForTesting`, which is why I moved it to the `internal` package. I felt like moving the other ones to the `internal` package would add too many symbols to it. Please let me know if you feel strongly about this. Thanks.

Comment (dfawley) [reply xds/xds.go:93]: IMO it's better for them to be in `internal` than visible in our external godoc.  I should have originally done `NewXDSResolver...` that way in hindsight.  I imagine it shouldn't be that hard to move these, but if it is, then you can leave them because they're also expected to be temporary.

Also, would this work??

```go
package internal

var NewXDSResolverBlah interface{}  // as always

.....

package internal/xds

func NewXDSResolverBlah(args) returns {
	// optionally handle `nil`??  Or can we actually import the thing that sets it from here??  Seems plausible.
	return internal.NewXDSResolverBlah(func(args)returns)(args)
}
```

That way we don't have to type cast the functions whenever we use them.  This could be a follow-up cleanup, and we could even do something similar with all the other stuff in `internal` that needs to be type cast every time we use it.

Comment (dfawley) [reply xds/xds.go:93]: > Or can we actually import the thing that sets it from here??  

I guess in the xds cases, "no", because of the visibility problem.  But in the other cases (e.g. ParseServiceConfig) then I think we could.

Comment (easwars) [reply xds/xds.go:93]: Moved everything to `internal` package. Thanks.

**Review by dfawley — APPROVED** [review]
